src/timestamp_generator.hpp

```diff
@@ -61,16 +61,24 @@ class ServerSideTimestampGenerator : public TimestampGenerator {
 
 class MonotonicTimestampGenerator : public TimestampGenerator {
 public:
-  MonotonicTimestampGenerator()
+  MonotonicTimestampGenerator(int64_t warning_threshold_us = 1000000,
+                              int64_t warning_interval_ms = 1000)
     : TimestampGenerator(MONOTONIC)
-    , last_(0) { }
+    , last_(0)
+    , last_warning_(0)
+    , warning_threshold_us_(warning_threshold_us)
+    , warning_interval_ms_(warning_interval_ms) { }
 
   virtual int64_t next();
 
 private:
   int64_t compute_next(int64_t last);
 
   Atomic<int64_t> last_;
+  Atomic<int64_t> last_warning_;
+
+  const int64_t warning_threshold_us_;
+  const int64_t warning_interval_ms_;
```

Comment:
No, it doesn't work like that. Try compiling this:  https://gist.github.com/mpenick/7878d5d328bbb9aa74a1f7dd125ebe15

```
$ clang++ -Wall -Wextra -pedantic unsigned.cpp
unsigned.cpp:5:30: warning: 'long long' is an extension when C99 mode is not enabled [-pedantic,-Wlong-long]
  printf("%llu\n", (unsigned long long)value);
                             ^
1 warning generated.
```

A warning is not generated for the negative value.